rosidl_runtime_cpp/include/rosidl_runtime_cpp/vector.hpp

```diff
@@ -0,0 +1,223 @@
+  }
+
+  operator bool &() {return value_;}
+
+  operator const bool &() const {return value_;}
+
+  bool * operator&() {return &value_;}  // NOLINT(runtime/operator)
+
+private:
+  bool value_;
+};
+
+static_assert(
+  std::is_standard_layout<BoolWrapper>::value,
+  "BoolWrapper must be a standard layout type "
+  "to afford reinterpret casts to bool.");
+
+}  // namespace internal
+
+/// A container based on std::vector with its own specializations.
+/**
+ * This container transparently interacts with std::vector<T> instances.
+ */
+template<typename T, typename AllocatorT = std::allocator<T>>
+class Vector : public std::vector<T, AllocatorT>
+{
+  using Base = std::vector<T, AllocatorT>;
+
+public:
+  using typename Base::value_type;
+  using typename Base::pointer;
+  using typename Base::const_pointer;
+  using typename Base::reference;
+  using typename Base::const_reference;
+  using typename Base::iterator;
+  using typename Base::const_iterator;
+  using typename Base::const_reverse_iterator;
+  using typename Base::reverse_iterator;
+  using typename Base::size_type;
+  using typename Base::difference_type;
+  using typename Base::allocator_type;
+
+  using Base::vector;
+
+  // Implicit "copy" and "move" constructors from std::vector<T> instances
+  Vector(const std::vector<T, AllocatorT> & other)  // NOLINT(runtime/explicit)
+  : Base(other) {}
+
+  Vector(std::vector<T, AllocatorT> && other)  // NOLINT(runtime/explicit)
+  : Base(other) {}
+};
+
+template<typename T, typename AllocatorT>
+inline void
+swap(Vector<T, AllocatorT> & x, Vector<T, AllocatorT> & y)
+{
+  x.swap(y);
+}
+
+/// Specialization for a boolean vector with no space optimizations.
+/**
+ * This container transparently interacts with std::vector<bool>
+ * instances while avoiding all the quirks in the latter.
+ */
+template<typename AllocatorT>
+class Vector<bool, AllocatorT>: public std::vector<
+    internal::BoolWrapper,
+    typename std::allocator_traits<AllocatorT>::
+    template rebind_alloc<internal::BoolWrapper>>
+{
+  using Base = std::vector<
+    internal::BoolWrapper,
+    typename std::allocator_traits<AllocatorT>::
+    template rebind_alloc<internal::BoolWrapper>>;
+
+public:
+  using value_type = bool;
+  using reference = bool &;
+  using const_reference = const bool &;
+  using pointer = bool *;
+  using const_pointer = const bool *;
+  using typename Base::iterator;
+  using typename Base::const_iterator;
+  using typename Base::reverse_iterator;
+  using typename Base::const_reverse_iterator;
+  using typename Base::size_type;
+  using typename Base::difference_type;
+  using typename Base::allocator_type;
+
+  using Base::vector;
+
+  // Implicit "copy" and "move" constructors from std::vector<bool> instances
+  Vector(const std::vector<bool, AllocatorT> & other)  // NOLINT(runtime/explicit)
+  : Base(other.begin(), other.end())
+  {
+  }
+
+  Vector(std::vector<bool, AllocatorT> && other)  // NOLINT(runtime/explicit)
```

Comment:
I think we should keep this implicit conversions, to make the transition smooth, but they are really bad performance wise.
IMHO, things that are costly should only be possible in an explicit way.

So I think we should add the implicit conversions with a deprecatd message, and provide explicit alternatives (e.g. a static method `from_std_vector_bool()`).
I don't mind much though.

Reply:
> I think we should keep this implicit conversions, to make the transition smooth, but they are really bad performance wise.

Yeah, I didn't want to cause breakage downstream but you're right.

> So I think we should add the implicit conversions with a deprecatd message, and provide explicit alternatives

That sounds good to me. Some `Vector::assign` overload and/or making constructors explicit would do too.